System/Posix/User.hsc

```diff
@@ -67,6 +68,108 @@ import System.IO.Error
 data {-# CTYPE "struct passwd" #-} CPasswd
 data {-# CTYPE "struct group"  #-} CGroup
 
+-- -----------------------------------------------------------------------------
+-- Thread safety of passwd/group database access APIs:
+--
+-- All supported unix platforms have @get(pw|gr)(nam|[ug]id)_r(3)@, which
+-- store the result in a caller provided buffer, which solves the most
+-- immediate thread-safety issues.
+--
+-- Things are more complicated for getpwent(3) and getgrent(3).
```

Comment:
Is it feasible to link correspondent man pages and sources for paragraphs below?

Reply:
Actually, looking closely at the FreeBSD code, the issues is actually an unintended side-effect (bug) in the "compat" passwd nsswitch API. The code for compat resets the global iterator key (`st->keynum`) also when doing a [single lookup](https://gitlab.unc.edu.ar/nicolaspapp/freebsd/-/blob/stable/12/lib/libc/gen/getpwent.c#L1774-1783) and then uses it as the [loop control variable](https://gitlab.unc.edu.ar/nicolaspapp/freebsd/-/blob/stable/12/lib/libc/gen/getpwent.c#L1825).

Changing `nsswitch` from `compat` to `files` solves the problem. So this is not a FreeBSD feature, but rather a long-standing bug. I can file a bug report and fix for FreeBSD, but it will still take a while for this to land on user systems. So I think we still need a work-around. :-(

Thoughts?

Reply:
Yeah, in the meantime a workaround is needed :(

Reply:
This is the fix:
```C
--- lib/libc/gen/getpwent.c     (revision 367119)
+++ lib/libc/gen/getpwent.c     (working copy)
@@ -1726,6 +1726,7 @@
 {
        char                     keybuf[MAXLOGNAME + 1];
        DBT                      key, entry;
+       pwkeynum                 keynum;
        struct compat_state     *st;
        enum nss_lookup_type     how;
        const char              *name;
@@ -1776,9 +1777,10 @@  
                        rv = NS_NOTFOUND;
                        goto fin;
                }
+               keynum = st->keynum;
                stayopen = 1;
        } else {
-               st->keynum = 0;
+               keynum = 0;
                stayopen = st->stayopen;
        }
 docompat:
@@ -1822,13 +1824,13 @@
        }
        key.data = keybuf;
        rv = NS_NOTFOUND;
-       while (st->keynum >= 0) {
-               st->keynum++;
+       while (keynum >= 0) {
+               keynum++;
                if (st->version < _PWD_CURRENT_VERSION) {
-                       memcpy(&keybuf[1], &st->keynum, sizeof(st->keynum));
-                       key.size = sizeof(st->keynum) + 1;
+                       memcpy(&keybuf[1], &keynum, sizeof(keynum));
+                       key.size = sizeof(keynum) + 1;
                } else {
-                       store = htonl(st->keynum);
+                       store = htonl(keynum);
                        memcpy(&keybuf[1], &store, sizeof(store));
                        key.size = sizeof(store) + 1;
                }
@@ -1839,7 +1841,7 @@   
                        rv = NS_UNAVAIL;
                        goto fin;
                } else if (rv == 1) {
-                       st->keynum = -1;
+                       keynum = -1;
                        rv = NS_NOTFOUND;
                        goto fin;
                }
@@ -1924,6 +1926,8 @@   
                        break;
        }
 fin:
+       if (how == nss_lt_all)
+               st->keynum = keynum;
        if (st->db != NULL && !stayopen) {
                (void)st->db->close(st->db);
                st->db = NULL;
```

Reply:
I filed a bug report: https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=252094
The patch is only for the `getwent()` issue, the `getgrent(3)` issue requires a somewhat more invasive fix, but still looks like a bug.

Reply:
FWIW, the getrennt(3) issue will also be addressed in a future FreeBSD release. But given the longevity of existing deployments, it'll be some years before we can assume the issue is no longer a concern.